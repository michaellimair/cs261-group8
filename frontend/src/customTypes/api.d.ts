type ApiBadRequestErrorKeys<T> = keyof T | 'non_field_errors';

export type IApiBadRequestErrorData<U> = {
  [key in ApiBadRequestErrorKeys<U>]: string[];
};
