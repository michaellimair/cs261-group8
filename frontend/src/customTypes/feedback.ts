import { IUser } from './auth';

export enum FeedbackType {
  BUG = 'bug',
  IMPROVEMENT = 'improvement',
  QUESTION = 'question',
}

export interface IFeedbackReply {
  id: number;
  content: string;
  created: Date;
  updated: Date;
  admin: IUser;
}

export interface IFeedback {
  id: number;
  type: FeedbackType;
  content: string;
  created: Date;
  updated: Date;
  reply: null | IFeedbackReply;
}
