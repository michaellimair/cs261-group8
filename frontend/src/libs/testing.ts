export const tick = (ms: number) => new Promise((resolve) => {
  setTimeout(resolve, ms);
});
