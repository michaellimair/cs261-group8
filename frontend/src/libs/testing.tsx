import { FC } from 'react';

import {
  MemoryRouter, Route, Routes,
} from 'react-router-dom';

export const tick = (ms: number) => new Promise((resolve) => {
  setTimeout(resolve, ms);
});

export const QueryRouterWrapper: FC = ({ children }) => (
  <MemoryRouter initialEntries={['/']}>
    <Routes>
      <Route path="*" element={children} />
    </Routes>
  </MemoryRouter>
);
