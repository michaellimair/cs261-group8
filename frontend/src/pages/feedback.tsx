import { FC } from 'react';

const FeedbackPage: FC = () => (
  <div>
    hello
  </div>
);

export default FeedbackPage;
