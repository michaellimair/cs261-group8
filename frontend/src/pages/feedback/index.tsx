import { FC } from 'react';
import { useTranslation } from 'react-i18next';
import { Heading, HStack, Stack } from '@chakra-ui/react';
import FeedbackList from 'components/Feedback/FeedbackList';
import CreateFeedbackLink from 'components/Feedback/CreateFeedbackLink';
import { httpClient } from 'api';
import { useQuery } from 'react-query';

const FeedbackPage: FC = () => {
  const { t } = useTranslation();
  const { data, isLoading } = useQuery(
    ['feedback', 'list'],
    () => httpClient.feedback.listFeedback(),
  );

  return (
    <Stack>
      <HStack justifyContent="space-between" alignItems="center">
        <Heading as="h2" size="md">
          {t('dashboard.feedback.your_feedbacks')}
        </Heading>
        <CreateFeedbackLink />
      </HStack>
      <FeedbackList isLoading={isLoading} feedbacks={data} />
    </Stack>
  );
};

export default FeedbackPage;
