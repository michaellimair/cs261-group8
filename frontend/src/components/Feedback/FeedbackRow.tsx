import {
  Stack, Text, Button, Box, useColorModeValue, VStack, HStack,
} from '@chakra-ui/react';
import { IFeedback } from 'customTypes/feedback';
import { FC } from 'react';
import { useTranslation } from 'react-i18next';
import FeedbackIcon from './FeedbackIcon';

interface IFeedbackRowProps {
  feedback: IFeedback;
}

const FeedbackRow: FC<IFeedbackRowProps> = ({
  feedback: {
    title, content, created, modified, type,
  },
}) => {
  const { t } = useTranslation();

  return (
    <Box p="4" boxShadow="lg" my="4" borderRadius="md" bg={useColorModeValue('white', 'gray.700')}>
      <Stack direction="row" alignItems="center" mb="2">
        <Text fontWeight="semibold" fontSize="xl">{title}</Text>
        <FeedbackIcon type={type} />
      </Stack>
      <VStack
        my="2"
        direction={{ base: 'column', md: 'row' }}
        alignItems="flex-start"
        justifyContent="space-between"
      >
        <Text fontSize={{ base: 'md' }} whiteSpace="pre-wrap">
          {content}
        </Text>
      </VStack>
      <Stack direction={{ base: 'column', lg: 'row' }} justifyContent="space-between" mt="4">
        <VStack
          alignItems="flex-start"
        >
          <Text fontSize={{ base: 'xs' }}>
            {t('created', { created })}
          </Text>
          {modified && (
          <Text fontSize={{ base: 'xs' }}>
            {t('modified', { modified })}
          </Text>
          )}
        </VStack>
        <HStack>
          <Button colorScheme="green">
            View
          </Button>
          <Button colorScheme="blue">
            Edit
          </Button>
          <Button colorScheme="red">
            Delete
          </Button>
        </HStack>
      </Stack>
    </Box>
  );
};

export default FeedbackRow;
