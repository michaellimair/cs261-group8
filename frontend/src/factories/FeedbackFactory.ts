/* eslint-disable class-methods-use-this */
import faker from '@faker-js/faker';
import { FeedbackType, IFeedback } from 'customTypes/feedback';
import { merge } from 'lodash';
import FeedbackReplyFactory from './FeedbackReplyFactory';

class FeedbackFactory {
  create = (feedback?: Partial<IFeedback>): IFeedback => {
    const replyFactory = new FeedbackReplyFactory();
    return merge({
      id: faker.datatype.number(),
      title: faker.lorem.lines(1),
      content: faker.lorem.paragraph(3),
      type: FeedbackType.BUG,
      created: new Date(),
      modified: new Date(),
      reply: replyFactory.create(),
    }, feedback);
  };

  createMany = (count: number) => Array(count).fill(null).map((_, i) => this.create({ id: i }));
}

export default FeedbackFactory;
