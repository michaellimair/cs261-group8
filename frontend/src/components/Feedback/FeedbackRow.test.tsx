import { render } from '@testing-library/react';
import { IFeedback } from 'customTypes/feedback';
import FeedbackFactory from 'factories/FeedbackFactory';
import { QueryRouterWrapper } from 'libs/testing';
import FeedbackRow from './FeedbackRow';

describe('components/Feedback/FeedbackRow', () => {
  const feedbackFactory = new FeedbackFactory();
  let feedback: IFeedback;

  beforeEach(() => {
    feedback = feedbackFactory.create();
  });

  it('renders correctly', () => {
    const result = render(
      <QueryRouterWrapper>
        <FeedbackRow feedback={feedback} />
      </QueryRouterWrapper>,
    );

    expect(result).toMatchSnapshot();
  });
});
