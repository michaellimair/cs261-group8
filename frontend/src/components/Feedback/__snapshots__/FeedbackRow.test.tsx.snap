// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Feedback/FeedbackRow renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div
      id="chakra-toast-portal"
    >
      <ul
        id="chakra-toast-manager-top"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        id="chakra-toast-manager-top-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-top-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-bottom-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-bottom"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        id="chakra-toast-manager-bottom-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
    </div>
    <div>
      <div
        class="css-jhpxgx"
      >
        <div
          class="chakra-stack css-1u8k0pa"
        >
          <p
            class="chakra-text css-d3lk8w"
          >
            Porro nulla id vero perspiciatis nulla nihil.
          </p>
          <span
            class="css-1qm7lvz"
          >
            <svg
              class="chakra-icon css-1azbes1"
              enable-background="new 0 0 48 48"
              fill="currentColor"
              focusable="false"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              version="1"
              viewBox="0 0 48 48"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M21.2,44.8l-18-18c-1.6-1.6-1.6-4.1,0-5.7l18-18c1.6-1.6,4.1-1.6,5.7,0l18,18c1.6,1.6,1.6,4.1,0,5.7l-18,18 C25.3,46.4,22.7,46.4,21.2,44.8z"
                fill="#F44336"
              />
              <path
                d="M21.6,32.7c0-0.3,0.1-0.6,0.2-0.9c0.1-0.3,0.3-0.5,0.5-0.7c0.2-0.2,0.5-0.4,0.8-0.5s0.6-0.2,1-0.2 s0.7,0.1,1,0.2c0.3,0.1,0.6,0.3,0.8,0.5c0.2,0.2,0.4,0.4,0.5,0.7c0.1,0.3,0.2,0.6,0.2,0.9s-0.1,0.6-0.2,0.9s-0.3,0.5-0.5,0.7 c-0.2,0.2-0.5,0.4-0.8,0.5c-0.3,0.1-0.6,0.2-1,0.2s-0.7-0.1-1-0.2s-0.5-0.3-0.8-0.5c-0.2-0.2-0.4-0.4-0.5-0.7S21.6,33.1,21.6,32.7z M25.8,28.1h-3.6L21.7,13h4.6L25.8,28.1z"
                fill="#fff"
              />
            </svg>
            <span
              class="css-8uhtka"
            >
              dashboard.feedback.bug
            </span>
          </span>
        </div>
        <div
          class="chakra-stack css-1dmv4b0"
        >
          <p
            class="chakra-text css-1k6unk6"
          >
            Eum iusto ullam saepe inventore sapiente enim eum. Autem vel non deleniti similique reprehenderit. Quo quasi laboriosam dicta est voluptatem consequatur fugiat a omnis. Accusamus accusamus reiciendis rem. Et praesentium occaecati dolor soluta sed quisquam rerum culpa.
          </p>
        </div>
        <div
          class="chakra-stack css-993yte"
        >
          <div
            class="chakra-stack css-1h4ws66"
          >
            <p
              class="chakra-text css-1859sd4"
            >
              created
            </p>
            <p
              class="chakra-text css-1859sd4"
            >
              modified
            </p>
          </div>
          <div
            class="chakra-stack css-84zodg"
          >
            <a
              class="chakra-link chakra-button css-ogxy0s"
              href="/54881"
            >
              view
            </a>
            <a
              class="chakra-link chakra-button css-ogxy0s"
              href="/54881/edit"
            >
              edit
            </a>
            <button
              class="chakra-button css-5c2176"
              type="button"
            >
              delete
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-jhpxgx"
    >
      <div
        class="chakra-stack css-1u8k0pa"
      >
        <p
          class="chakra-text css-d3lk8w"
        >
          Porro nulla id vero perspiciatis nulla nihil.
        </p>
        <span
          class="css-1qm7lvz"
        >
          <svg
            class="chakra-icon css-1azbes1"
            enable-background="new 0 0 48 48"
            fill="currentColor"
            focusable="false"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            version="1"
            viewBox="0 0 48 48"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21.2,44.8l-18-18c-1.6-1.6-1.6-4.1,0-5.7l18-18c1.6-1.6,4.1-1.6,5.7,0l18,18c1.6,1.6,1.6,4.1,0,5.7l-18,18 C25.3,46.4,22.7,46.4,21.2,44.8z"
              fill="#F44336"
            />
            <path
              d="M21.6,32.7c0-0.3,0.1-0.6,0.2-0.9c0.1-0.3,0.3-0.5,0.5-0.7c0.2-0.2,0.5-0.4,0.8-0.5s0.6-0.2,1-0.2 s0.7,0.1,1,0.2c0.3,0.1,0.6,0.3,0.8,0.5c0.2,0.2,0.4,0.4,0.5,0.7c0.1,0.3,0.2,0.6,0.2,0.9s-0.1,0.6-0.2,0.9s-0.3,0.5-0.5,0.7 c-0.2,0.2-0.5,0.4-0.8,0.5c-0.3,0.1-0.6,0.2-1,0.2s-0.7-0.1-1-0.2s-0.5-0.3-0.8-0.5c-0.2-0.2-0.4-0.4-0.5-0.7S21.6,33.1,21.6,32.7z M25.8,28.1h-3.6L21.7,13h4.6L25.8,28.1z"
              fill="#fff"
            />
          </svg>
          <span
            class="css-8uhtka"
          >
            dashboard.feedback.bug
          </span>
        </span>
      </div>
      <div
        class="chakra-stack css-1dmv4b0"
      >
        <p
          class="chakra-text css-1k6unk6"
        >
          Eum iusto ullam saepe inventore sapiente enim eum. Autem vel non deleniti similique reprehenderit. Quo quasi laboriosam dicta est voluptatem consequatur fugiat a omnis. Accusamus accusamus reiciendis rem. Et praesentium occaecati dolor soluta sed quisquam rerum culpa.
        </p>
      </div>
      <div
        class="chakra-stack css-993yte"
      >
        <div
          class="chakra-stack css-1h4ws66"
        >
          <p
            class="chakra-text css-1859sd4"
          >
            created
          </p>
          <p
            class="chakra-text css-1859sd4"
          >
            modified
          </p>
        </div>
        <div
          class="chakra-stack css-84zodg"
        >
          <a
            class="chakra-link chakra-button css-ogxy0s"
            href="/54881"
          >
            view
          </a>
          <a
            class="chakra-link chakra-button css-ogxy0s"
            href="/54881/edit"
          >
            edit
          </a>
          <button
            class="chakra-button css-5c2176"
            type="button"
          >
            delete
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
